### PR TITLE
8317809: Insertion of free code blobs into code cache can be very slow during class unloading

### PR DESCRIPTION
**Pull request by tschatzl**
Insert code blobs in a sorted fashion to exploit the finger-optimization when adding, making this procedure O(n) instead of O(n^2)

Introduces a globally available ClassUnloadingContext that contains common methods pertaining to class and code unloading. GCs may use it to efficiently manage unlinked class loader datas and nmethods to allow use of common methods (unlink/merge).

The steps typically are registering a new to be unlinked CLD/nmethod, and then purge its memory later. STW collectors perform this work in one big chunk taking the CodeCache_lock, for the entire duration, while concurrent collectors lock/unlock for every insertion to allow for concurrent users for the lock to progress.

Some care has been taken to stay consistent with an "unloading = unlinking + purge" scheme; however particularly the existing CLD handling API (still) mixes unlinking and purging in its CLD::unload() call. To simplify this change that is mostly geared towards separating nmethod unlinking from purging, to make code blob freeing O(n) instead of O(n^2).

Upcoming changes will
* separate nmethod unregistering from nmethod purging to allow doing that in bulk (for the STW collectors); that can significantly reduce code purging time for the STW collectors.
* better name the second stage of unlinking (called "cleaning" throughout, e.g. the work done in `G1CollectedHeap::complete_cleaning`)
* untangle CLD unlinking and what's called "cleaning" now to allow moving more stuff into the second unlinking stage for better parallelism
* G1: move some significant tasks from the remark pause to concurrent (unregistering nmethods, freeing code blobs and cld/metaspace purging)
* Maybe move Serial/Parallel GC metaspace purging closer to other unlinking/purging code to keep things local and allow easier logging.

Please also first looking into the (small) PR this depends on.

The crash on linux-x86 is fixed by PR#16766 which I split out for quicker reviews.

Testing: tier1-7

Thanks,
  Thomas

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8317809](https://bugs.openjdk.org/browse/JDK-8317809): Insertion of free code blobs into code cache can be very slow during class unloading (**Enhancement** - P4)

### Reviewers
 * [Ivan Walulya](https://openjdk.org/census#iwalulya) (@walulyai - **Reviewer**)
 * [Albert Mingkun Yang](https://openjdk.org/census#ayang) (@albertnetymk - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16759/head:pull/16759` \
`$ git checkout pull/16759`

Update a local copy of the PR: \
`$ git checkout pull/16759` \
`$ git pull https://git.openjdk.org/jdk.git pull/16759/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16759`

View PR using the GUI difftool: \
`$ git pr show -t 16759`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16759.diff">https://git.openjdk.org/jdk/pull/16759.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16759#issuecomment-1821221735)

### CODE REVIEW

**Comment by tschatzl** [comment]
I added some explanation of why there is a class hierarchy for `ClassUnloadingContext` in the description (and some further background).

File: src/hotspot/share/classfile/classLoaderData.cpp [hunk 1]
```diff
@@ -599,7 +599,7 @@ void ClassLoaderData::unload() {
   free_deallocate_list_C_heap_structures();
 
   // Clean up class dependencies and tell serviceability tools
-  // these classes are unloading.  Must be called
+  // these classes are unloading.  This must be called
```
Comment (offamitkumar) [inline src/hotspot/share/classfile/classLoaderData.cpp:602]:
```suggestion
  // these classes are unloading. This must be called
```

Comment (tschatzl) [reply src/hotspot/share/classfile/classLoaderData.cpp:602]: Hotspot code style allows comments with both two spaces and one space between sentences in a paragraph. I did not want to change the style there.

File: src/hotspot/share/gc/shared/classUnloadingContext.hpp [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+  void purge_and_free_nmethods() {
+    purge_nmethods();
+    free_code_blobs();
+  }
+};
+
+class DefaultClassUnloadingContext : public ClassUnloadingContext {
+  ClassLoaderData* volatile _cld_head;
+
+  uint _num_nmethod_unlink_workers;
```
Comment (walulyai) [inline src/hotspot/share/gc/shared/classUnloadingContext.hpp:66]: probably better to set `_num_nmethod_unlink_workers;` as const, the destructor depends on this being const through the lifetime of the object.

File: src/hotspot/share/gc/shared/classUnloadingContext.hpp [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+  void classes_unloading_do(void f(Klass* const)) override;
+
+  // Register unloading nmethods, potentially in parallel.
+  void register_unlinked_nmethod(nmethod* nm) override;
+  void purge_nmethods() override;
+  void free_code_blobs() override;
+};
+
+#endif // SHARE_GC_SHARED_CLASSUNLOADINGCONTEXT_HPP
```
Comment (walulyai) [inline src/hotspot/share/gc/shared/classUnloadingContext.hpp:95]: trailing space warning*

File: src/hotspot/share/gc/shared/classUnloadingContext.hpp [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+  virtual void purge_nmethods() = 0;
+  virtual void free_code_blobs() = 0;
+
+  void purge_and_free_nmethods() {
+    purge_nmethods();
+    free_code_blobs();
+  }
+};
+
+class DefaultClassUnloadingContext : public ClassUnloadingContext {
```
Comment (albertnetymk) [inline src/hotspot/share/gc/shared/classUnloadingContext.hpp:63]: I don't understand why they need to be two classes, even after reading "These are the reason for the class hierarchy for...". The reference to future/other PR(s) in the description doesn't really help -- it's unclear what is *necessary* for the current PR and what is preparation for future PR(s).

File: src/hotspot/share/gc/shared/classUnloadingContext.hpp (hunk 3 above)
Comment (albertnetymk) [inline src/hotspot/share/gc/shared/classUnloadingContext.hpp:95]: Seems missing a newline.

File: src/hotspot/share/gc/shared/classUnloadingContext.cpp [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+
+void DefaultClassUnloadingContext::free_code_blobs() {
+  assert(_context != nullptr, "no context set");
+
+  // Sort nmethods before freeing to benefit from optimizations. If Nmethods were
+  // collected in parallel, use a new temporary buffer for the result, otherwise
+  // sort in-place.
+  NMethodSet* nmethod_set = nullptr;
+
+  bool is_parallel = _num_nmethod_unlink_workers != 1;
```
Comment (walulyai) [inline src/hotspot/share/gc/shared/classUnloadingContext.cpp:138]:
```suggestion
  bool is_parallel = _num_nmethod_unlink_workers > 1;
```
Seems more intuitive.

File: src/hotspot/share/gc/shared/classUnloadingContext.hpp (hunk 4 above)
Comment (tschatzl) [reply src/hotspot/share/gc/shared/classUnloadingContext.hpp]: As explained in the now removed original description:


>These are the reason for the class hierarchy for `ClassUnloadingContext`: the goal is to ultimately have about this phasing (for G1):
>1. collect all dead CLDs, using the `register_unloading_class_loader_data` method *only*
>2. parallelize the stuff in `ClassLoaderData::unload()` in one way or another, adding them to the `complete_cleaning` >(parallel) phase.
>
>Particularly the split of `SystemDictionary::do_unloading` into "only" traversing the CLDs to find the dead ones and then in parallel process them in 2. above warrants a separate `ClassUnloadingContext` (to facilitate parallelism). 


I.e. the non-parallelized case does not need the necessary data structure complications and helper methods to do efficient parallel iteration. However as mentioned I removed the class hierarchy for this change as it's unnecessary for now; let's discuss this hierarchy separately.

**Review by albertnetymk — APPROVED** [review]
Only one minor & subjective comment.

**Review by walulyai — APPROVED** [review]

File: src/hotspot/share/gc/shared/classUnloadingContext.cpp [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+  }
+}
+
+void ClassUnloadingContext::classes_unloading_do(void f(Klass* const)) {
+  assert_locked_or_safepoint(ClassLoaderDataGraph_lock);
+  for (ClassLoaderData* cld = _cld_head; cld != nullptr; cld = cld->unloading_next()) {
+    assert(cld->is_unloading(), "invariant");
+    cld->classes_do(f);
+  }
+}
```
Comment (coleenp) [inline src/hotspot/share/gc/shared/classUnloadingContext.cpp:91]: I don't understand why CLDG specific methods were moved here.  They should be unaware of nmethod purging. and these 4 methods don't have any nmethod purging in them either and are specific to the CLDG implementation.

Comment (tschatzl) [reply src/hotspot/share/gc/shared/classUnloadingContext.cpp:91]: The idea is to have GC take control how the unloading CLDs are stored/which data structure it is going to use to manage them to ultimately allow more control about class unloading for parallelization.

Which on the one hand makes pauses shorter (for stw collectors), and on the other hand decreases the time the CLDG_lock is held (not sure it is nice that the concurrent collectors currently may hold that one for ~100ms in my test...).

I believe having the linked list of unloading CLDs embedded in the CLDs for use by the GC not only seems wrong (i.e. it's a GC data structure located in runtime code) but is also very limiting (need to have one for all, fixed singly linked list).

This change moves knowledge of how unloading CLDs are managed to GC area - runtime code just tells GC that a particular CLD is unloading.
(Currently the `ClassUnloadingContext` also calls the `unload` method during registration to keep current functionality, but the plan is to separate the step of registration and actual unloading to allow custom handling of the second part; the registering, although it's still walking a singly linked list, is comparatively fast).

These four methods provide a thin abstraction over the CLDs that are unloading (that runtime doesn't need and should not worry about imo).

With that in place it is possible to slice the actual unloading work into phases according to dependencies (depending on GC if desired), potentially overlapping with other existing phases in collectors already allowing that (e.g. the parallel code unloading, but that is only an implementation detail to reduce overall parallel phases), or even moving some of that work sometime else (the `CLD::unload()` method unfortunately currently may do some memory freeing too).

However most time is spent in notifying various components which can be parallelized (at least parallelize the different types of notifications).

**Pull request closed without merging**